xrdp/xrdp_mm.c

```diff
@@ -2314,26 +2326,24 @@ xrdp_mm_check_wait_objs(struct xrdp_mm *self)
 int APP_CC
 xrdp_mm_frame_ack(struct xrdp_mm *self, int frame_id)
 {
-    int ex;
+    struct xrdp_encoder *encoder;
 
-    LLOGLN(10, ("xrdp_mm_frame_ack: incoming %d, client %d, server %d", frame_id,
-        self->encoder->frame_id_client, self->encoder->frame_id_server));
-    self->encoder->frame_id_client = frame_id;
+    LLOGLN(10, ("xrdp_mm_frame_ack:"));
     if (self->wm->client_info->use_frame_acks == 0)
     {
         return 1;
     }
-    ex = self->wm->client_info->max_unacknowledged_frame_count;
-    /* make sure we won't have too many in-flight frames */
-    if (self->encoder->frame_id_client + ex >= self->encoder->frame_id_server)
+    encoder = self->encoder;
+    LLOGLN(10, ("xrdp_mm_frame_ack: incoming %d, client %d, server %d",
+           frame_id, encoder->frame_id_client, encoder->frame_id_server));
+    if (frame_id < 0)
     {
-        if (self->encoder->frame_id_server > self->encoder->frame_id_server_sent)
-        {
-            LLOGLN(10, ("xrdp_mm_frame_ack: frame_id_server %d", self->encoder->frame_id_server));
-            self->encoder->frame_id_server_sent = self->encoder->frame_id_server;
-            self->mod->mod_frame_ack(self->mod, 0, self->encoder->frame_id_server);
-        }
+        /* if frame_id == -1 ack all sent frames */
+        encoder->frame_id_client = encoder->frame_id_server;
```

Comment:
If -1 has that special property, let's make a define for it. I'm not sure if other negative numbers should have that properly. The code and the comment differ here.